### PR TITLE
Display Asterisk on Unsaved Documents

### LINKED ISSUE #268: Display an asterisk next to document tabs and page title with unsaved work
Making changes to the document should display the asterisk. Performing undo back to the most-recently-saved ~~(at the time of writing, that means the beginning of document history since there is no saving support yet)~~ should remove the asterisk.

The page title should also update to show the name of the current document including any asterisk.

Related: #267

### PR DESCRIPTION
**Pull request by mfish33**
Closes #268

An asterisk will now display next to the document tab as well as the title when it is unsaved. Detection for an unsaved document is "ok" as described in the code comments. In order to have more accurate detection, there may need to be changed on the rust side as well.

In addition, I have moved wasm-pack to optional dependencies since on the m1 mac it will fail to install.

---
This change is [Reviewable](https://reviewable.io/reviews/graphiteeditor/graphite/392)

### CODE REVIEW

File: frontend/src/utilities/documents.ts
Comment (Keavon) [inline frontend/src/utilities/documents.ts:101]: Let's change it so the Rust code sends all this information instead of just the names. We can pair program this part so I can show you the Rust stuff.

File: frontend/src/utilities/documents.ts
Comment (Keavon) [inline frontend/src/utilities/documents.ts:118]: This is a nice attempt working with what you have in the browser, since it has very little "context" about the state of the application. Although I believe the only way to do this "correctly" would be to edit the Rust code which has the entire application state and contextual information. Let's pair program this so I can show you the Rust codebase and we can make this work based on any change that has occurred since the document was saved.

File: frontend/src/utilities/documents.ts
Comment (Keavon) [inline frontend/src/utilities/documents.ts:164]: Let's make this happen in the Rust code, since the frontend isn't supposed to modify state.

File: frontend/package.json
Comment (Keavon) [inline frontend/package.json:43]: Let's save this change for another PR, since it might have some unintended consequences. Such as: the change to `package-lock.json`, which might or might not be related to this change, has had 15.5k lines added to it for some reason. We should investigate this further. But for now, please discard the changes to `package-lock.json` and `package.json` right before we merge this PR.

**Review by Keavon — COMMENTED** [review]
Nice work! Let's do some pair programming/debugging to figure out the things mentioned in the comments.

**Comment by bjorn3** [comment]
If you revert all changes since the last save using ctrl+z or ctrl+shift+z does ot revert back to showing that it is already saved?

**Comment by mfish33** [comment]
The revised rust implementation works properly with undo and redo since it is now hooked into the history system. In addition, Eslint needed to be upgraded because Typescript 4.0 changed the way that tuple types are represented in the AST. As a consequence of upgrading, some rules became stricter and I updated the code in a few places to comply. Finally, the reason for the large package.json is that I am on a newer version of npm. Missed this when I ran npm install for the first time:
```bash
npm WARN old lockfile The package-lock.json file was created with an old version of npm,
npm WARN old lockfile so supplemental metadata must be fetched from the registry.
npm WARN old lockfile 
npm WARN old lockfile This is a one-time fix-up, please be patient...
```
@Keavon Do we want to upgrade this for the project?

File: frontend/package.json
Comment (Keavon) [reply frontend/package.json:43]: I went ahead and made #395 which, with some careful experimentation, upgrades everything that we can possibly upgrade except the things that are blocked on Vue CLI 5.0 which is in the release candidate stage right now. This also cleaned up some new linter errors and things. Some of this might be a duplication of your work, but it was better suited for another PR. If you made any improvements that I didn't include also in that commit, please feel free to include them. Sorry if this means you have some merge conflicts, I am more than happy to help you sort them out.

**Comment by Keavon** [comment]
> If you revert all changes since the last save using ctrl+z or ctrl+shift+z does ot revert back to showing that it is already saved?

@bjorn3 this works properly now, thanks.

**Comment by Keavon** [comment]
Since it's likely to be buried in the resolved code review comment [here](https://github.com/GraphiteEditor/Graphite/pull/392#discussion_r758176890), let me copy this comment I wrote into the main thread:

I went ahead and made #395 which, with some careful experimentation, upgrades everything that we can possibly upgrade except the things that are blocked on Vue CLI 5.0 which is in the release candidate stage right now. This also cleaned up some new linter errors and things. Some of this might be a duplication of your work, but it was better suited for another PR. If you made any improvements that I didn't include also in that commit, please feel free to include them. Sorry if this means you have some merge conflicts, I am more than happy to help you sort them out.

**Comment by Keavon** [comment]
@mfish33 we also assigned you to #267 which is a natural extension to this feature, and should be relatively small to tackle once this one is merged. Feel free to un-assign yourself if you'd prefer not to work on it. And of course, if you have any questions or want to pair program on it, I'm always very happy to help!

File: editor/src/document/document_file.rs
Comment (TrueDoctor) [inline editor/src/document/document_file.rs:359]: It feels a bit weird for these function to unconditionally push an update to the message queue, I'm wondering if there are situations where you would want to use the history operations without a message queue. But this probably helps to avoid errors

Comment (TrueDoctor) [reply editor/src/document/document_file.rs:359]: (but you could return the message as the Ok() part, then clippy is going to complain if you don't use the `result but you could still use the function without a message queue)`

File: editor/src/document/document_message_handler.rs
Comment (TrueDoctor) [inline editor/src/document/document_message_handler.rs:104]: @Keavon this sounds a bit off, shouldn't it be `was not found` or `is not in`?

File: editor/src/document/document_file.rs
Comment (TrueDoctor) [inline editor/src/document/document_file.rs:373]: The GetOpenDocumentsList Message sholud probably be renamed to UpdateOpenDocumentsList, righ?

File: editor/src/document/document_file.rs
Comment (mfish33) [reply editor/src/document/document_file.rs:394]: Is this expensive? calling:
```rs
graphene_document.hash()
```
results in just calling:
```rs
self.hasher.finish()
```
does hasher.finish have a time complexity greater than O(1)? This could be replaced with a counter that starts at 0 when saved and increments on a redo or a new change and then decrements on an undo. However, this moves the responsibility of maintaining the counter to several places inside document_file instead of a single source of truth. However, if you think that it will help performance I can change it.

File: editor/src/document/document_file.rs
Comment (mfish33) [reply editor/src/document/document_file.rs:359]: That's a good point. However, do we want to allow this function to be run without the message queue? Currently, we want to send the GetOpenDocumentsList message at all times, so it would lead to code duplication that if missed at a later time could create a hard-to-track down bug.

File: editor/src/document/document_message_handler.rs
Comment (mfish33) [reply editor/src/document/document_message_handler.rs:104]: Good catch

File: editor/src/document/document_file.rs
Comment (mfish33) [reply editor/src/document/document_file.rs:373]: changed

File: editor/src/document/document_file.rs
Comment (TrueDoctor) [reply editor/src/document/document_file.rs:394]: Oh, yeah was wondering what sort of hash function you were using… lets say there is a reason for the hashing function containing unsafe code, the "hash" is not "really" as hash, it is more of a hash with some randomnes sprinkled in.
I thought I had removed that impl when transitioning to the uuid based system.
The implementation you are currently using is not a performance concern, but well it kinda relies on a bit of undefined behaviour and `std::mem::transmute_copy`'s.
If there is enough value in the current system, we should keep it, otherwise we should probably replace it with something sensible.
The main issue with the current system is that hash(x) might not be equal to hash(x), meaning x == y does not imply hash(x) == hash(y)

Comment (TrueDoctor) [reply editor/src/document/document_file.rs:394]: You can also take a look at the hash implementation yourself and judge whether you can reconcile this with your conscience. Float are just a pain to deal with

File: editor/src/document/document_file.rs
Comment (TrueDoctor) [reply editor/src/document/document_file.rs]: Yeah its fine as it is we can change it should it become a problem

File: editor/src/document/document_file.rs
Comment (mfish33) [reply editor/src/document/document_file.rs:394]: I went to implement it as a counter, however, it becomes a problem since once a new change is made to the document after undoing needs to invalidate the counter permanently. There are ways that this can be done but it starts to feel like the incorrect solution. The hash I think is safe to use since I am never using it to compare two different documents. The documents saved in document_undo_history are used immutable so I am using the hash to just identify a single state of a document.

Comment (mfish33) [reply editor/src/document/document_file.rs:394]: It might make sense to change the name from hash to identifier. This would alleviate some of the concerns around the fact that x == y does not imply hash(x) == hash(y)

Comment (TrueDoctor) [reply editor/src/document/document_file.rs:394]: Yeah that's a good idea, I thought I had already done that and then crossed it off my mental todo list. If you want you can append the change to this pr, otherwise I'll create an issue for it.

Comment (mfish33) [reply editor/src/document/document_file.rs]: OK. I added those changes plus a note about the identifier to try and prevent misconceptions in the future.

**Review by TrueDoctor — APPROVED** [review]
Looks good from the rust side, the package.lock file is a bit concerning I thought that was fixed in master?

**Comment by Keavon** [comment]
> Looks good from the rust side, the package.lock file is a bit concerning I thought that was fixed in master?

Yeah, it still needs to be merged with those changes in master. Happy to help with that if you have any troubles @mfish33.

**Comment by TrueDoctor** [comment]
> Yeah, it still needs to be merged with those changes in master. Happy to help with that if you have any troubles @mfish33.

Is it an option to just remove the file and regenerate it?

**Comment by mfish33** [comment]
Just finished, just did it in two steps, because I did not want to manually make sure that package-lock.json was merged correctly since my fork had a different package.json

File: frontend/src/utilities/documents.ts
Comment (Keavon) [inline frontend/src/utilities/documents.ts:93]: This is left over from debugging

**Review by Keavon — APPROVED** [review]

**Pull request merged**
